---
title: "Partial Least Squares"
output:
  html_document:
    keep_md: true    
---

```{r setup-chunk, include=FALSE}
knitr::opts_chunk$set(dev = "png",
                      dpi = 900,
                      echo = FALSE,
                      cache = TRUE)
```


# Import Libraries
```{r}
library(NanoStringNCTools)
library(GeomxTools)

library(SpatialExperiment)
library(standR)
library(pls)

library(dplyr)
library(tidyverse)

library(ggplot2)
library(ggalluvial)
library(ggrepel)

#library(pls)
library(mixOmics)
```

#Load Data
```{r}
datadir <- "data/"
DCCFiles <- dir(datadir, pattern=".dcc$", full.names=TRUE)
PKCFiles <- unzip(zipfile = file.path(datadir,  "/Mm_R_NGS_WTA_v1.0.zip"))
SampleAnnotationFile <- file.path(datadir, "ModifiedWorksheet2.xlsx")

dataSet <-
  readNanoStringGeoMxSet(dccFiles = DCCFiles,
                                          pkcFiles = PKCFiles,
                                          phenoDataFile = SampleAnnotationFile,
                                          phenoDataSheet = "Sheet1",
                                          phenoDataDccColName = "Sample_ID",
                                          protocolDataColNames = c("Roi",
                                                                   "Aoi"))

dataSet <- dataSet[,pData(dataSet)$SlideName != "No Template Control"] #remove no template control

#make spatial experiment object
dataSet_target <- aggregateCounts(dataSet)
seDataSet <- as.SpatialExperiment(dataSet_target, normData = "exprs" ,forceRaw = TRUE)

assay(seDataSet, "counts") <- assay(seDataSet, "GeoMx")
assay(seDataSet, "logcounts") <- log2(assay(seDataSet, "GeoMx") + 1)
```

#QC and Normalization
```{r}
#Gene Level
dim(seDataSet)

seDataSet <- addPerROIQC(seDataSet, rm_genes=TRUE, sample_fraction = 0.9, min_count = 5)

dim(seDataSet)
metadata(seDataSet) |> names()


plotGeneQC(seDataSet, ordannots = "Segment", col = Segment, point_size = 2, top_n = 9)

#ROI Level
#On RLE plots:
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5798764/

plotROIQC(seDataSet, x_axis="Nuclei", y_axis = "lib_size",  x_lab = "Nuclei", x_threshold = 50, color = Segment)

plotROIQC(seDataSet, x_axis="Area", y_axis = "lib_size", x_lab = "Area", color = Segment)


plotRLExpr(seDataSet, ordannots = "Segment", color = Segment, assay = 2) #counts
plotRLExpr(seDataSet, ordannots = "Segment", color = Segment, assay = 3) #logcounts

seDataSet_tmm <- geomxNorm(seDataSet, method = "TMM")
tmmSubset <- seDataSet_tmm[, seDataSet_tmm$Segment == "PanCK+"]

plotRLExpr(seDataSet_tmm, ordannots = "Segment", color = Segment, assay = 3) #logcounts
```

#Import ROI summary data and do PLS
```{r}
flimParams <- read_csv("matlab scripts/roiSummaryTable.csv")

countsMatrix <- t(assay(tmmSubset, 'logcounts'))
colMatrx <- as.data.frame(colData(tmmSubset))

joined <- left_join(colMatrx, flimParams, by = "sample")

#for pls package:
plsrModelLTint <- plsr(cbind(joined$LT1mean, joined$LT2mean, joined$LT3mean, joined$IR13mean, joined$IR23mean) ~ countsMatrix, scale = TRUE)
plsrModelLT <- plsr(cbind(joined$LT2mean) ~ countsMatrix, scale = TRUE)

#for mixOmics:
  #http://mixomics.org/methods/spls/
flimParamsRaw <- subset(joined, select = c(LT2mean, LT3mean))
splsResult <- spls(countsMatrix, flimParamsRaw, mode = "regression", scale = TRUE, ncomp = 3, keepX = 50)
```
our problem is one of 'feature selection'
  -we should prefer a 'regression' approach because we expect a causual relationship between sets (transcripts -> fluorophores -> LT measurment) ??
-this only matters for (s)PLS after first component becauce the difference is in how the matricies are deflated  
https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-34

https://doi.org/10.1093/bib/bbl016
  -ordering of weights only has clear statistical meaining for first component, single regression (not our case)

sPCS  
https://doi.org/10.2202/1544-6115.1390

to look into:
https://github.com/gdurif/plsgenomics
http://mixomics.org/methods/spls/

review of selection methods:
https://doi.org/10.1016/j.chemolab.2012.07.010
#Investigate
```{r}
#summary(plsrModelLT)
#summary(plsrModelLTint)

splsResult$prop_expl_var

variableImportance <- as.data.frame(vip(splsResult)) %>% arrange(desc(comp1))
head(variableImportance, n = 50)
```

